### PR TITLE
feat(serve): status reactions, reaction answers, and attention nudges (1.1.0)

### PR DESCRIPTION
**Pull request by anaclumos**
## What

Slack UX improvements for `xx serve` (owner ask 2026-07-20):

- **Status reactions** on the turn's triggering message: hourglass while processing (parks/retries included; the id rides the activeTurn marker so recovered turns finish the lifecycle), settled to check (done) / x (failed; resume give-up too) / question (waiting on you). Failure beats attention; a drain-presumed-killed turn keeps its hourglass for the resume; finish+attention settles done. All reaction API failures are log-only: decoration never fails a turn.
- **need_attention**: new in-process MCP tool beside finish_thread; the ask-the-user skill instructs calling it in the asking turn. settleTurn persists an attention state on the thread record; any relayable reply clears it.
- **Reactions as answers**: an asked user's reaction relays as their answer turn when it passes all gates (attention pending, reactor in requesterIds, reaction postdates askedAt via raw event_ts, reacted message at/after the ask trigger). No pre-clear: handleTurn's consume step is the single clear point, so failed fetches / unlinked channels / drains leave the ask intact. Everything else folds into the next turn's prompt as an unmetered note, guarded by a fail-closed home-workspace check (users.info, cached) and a structural emoji-name allowlist.
- **Nudge**: 60s sweep posts one mention-tagging reminder per ask after 10 unanswered minutes; restart-safe via persisted nudgedAt; unlinked channels skipped.
- **Manifest**: adds `reactions:read`, `reactions:write`, and the `reaction_added` event.

## Operator action required

The live Slack app must have the updated manifest re-pasted and be reinstalled to the workspace, or reactions silently no-op (logged as `serve.reaction_error`).

## Verification

- `bun run typecheck` clean; `bun test` 428 pass / 0 fail (19 new tests).
- Pre-PR unhinted adversarial review (3 independent reviewers, 2 skeptics per finding): 6 confirmed findings, all fixed in the second commit; 1 refuted, 0 contested.
- Docs: serve.mdx + all 29 locales, AGENTS.md operational notes.

## Known limitations (for the owner, not blockers)

- Reaction-vs-ask ordering uses Slack event_ts vs the daemon's askedAt clock; skew misclassifies toward the safe side (note path, never auto-approval).
- The nudge fires once per ask by design (mention fatigue; the ask itself already tagged once).

---

> [!NOTE]
> **Medium Risk**
> Expands serve daemon behavior (reactions, inbound events, persisted thread state) and Slack OAuth scopes; reaction API failures are isolated log-only, with broad test coverage for gates and edge cases.
> 
> **Overview**
> **Slack `serve` UX (1.2.0):** turns now show progress on the **triggering message** via emoji reactions (hourglass → check / x / question), with special cases so drain-killed and usage-limit deferrals keep the hourglass until auto-resume.
> 
> Adds an in-process **`need_attention`** MCP tool (alongside `finish_thread`) so the model can flag threads waiting on a decision; the daemon persists **`attention`** on thread records, clears it when the user replies, and runs a **60s nudge sweep** (one @-mention after 10 minutes). **`bot.onReaction`** routes user emoji: qualified reactions from asked users become answer turns; others fold into the next prompt as context, with home-workspace verification and emoji allowlisting. Slack app manifest gains **`reactions:read`**, **`reactions:write`**, and **`reaction_added`** (reinstall required).
> 
> Docs and **`AGENTS.md`** are updated across locales; **`ask-the-user`** skill instructs calling `need_attention`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d42b4f4125ed47292e9c7a9a6ff843d2a1c95538. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Slack UX upgrades for `serve`: status reactions on the trigger, reaction-based answers while waiting, and a one-time nudge if the user goes quiet. Docs updated across locales; deferrals keep the hourglass, recovery keeps the original askers, unlinked reactions drop, unknown-recovery drops now settle their status, and reply-splitting notes are included (1.2.0).

- **New Features**
  - Status reactions: hourglass while a turn runs (parks/retries/deferrals), then check (done), x (failed), or question (waiting); recovered turns finish via the stored message id; unknown-recovery terminal drops now settle x and clear the hourglass; reaction API failures are log-only.
  - `need_attention` tool: marks the thread as waiting; any relayable reply clears it; recovery uses the original requesterIds for nudging and answer gating.
  - Reactions as answers: while waiting, an asked user’s reaction posted after the ask on or after the trigger becomes their answer; other reactions fold into the next prompt as an unmetered note with safety checks; during a drain/deferral, reactions take the note path; reactions in unlinked channels are dropped.

- **Migration**
  - Update the Slack app manifest to include `reactions:read`, `reactions:write`, and the `reaction_added` event, then reinstall the app; otherwise reaction-based features (status, answers, attention handling) will no-op.

<sup>Written for commit 506bd5ae74cb1c32d7aea7848e6da174eeadf6c8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/anaclumos/tokenmaxxing/pull/43?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: src/cli/serve.ts [hunk 1]
```diff
@@ -518,26 +603,64 @@ export function buildServeRuntime(seam: {
...
+    // predicate as runTurn's marker retention) keeps its hourglass - it will
+    // auto-resume next start, and a terminal x nothing ever removes would
+    // read a later successful resume as failed; and a finished thread settles
+    // as done even when the model also flagged attention, because the record
+    // deletion below makes the question mark unremovable forever.
+    const killedByDrain = draining && outcome.failed && !outcome.announcedDrop && !outcome.resultReceived;
+    if (input.messageId && !killedByDrain) {
+      const emoji = outcome.failed ? STATUS_EMOJI.failed : outcome.attention && !outcome.finish ? STATUS_EMOJI.attention : STATUS_EMOJI.done;
+      await setStatus({ threadId: thread.id, messageId: input.messageId, emoji, op: "add" });
+      await setStatus({ threadId: thread.id, messageId: input.messageId, emoji: STATUS_EMOJI.processing, op: "remove" });
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/cli/serve.ts:659]: Agreed, fixed in 91ab6b0: a usage-limit deferral now keeps its hourglass exactly like the drain-presumed-killed turn (same exception, same reason: the durable marker promises an auto-resume, and a terminal x nothing removes would read a later successful resume as failed). Pinned by test.

File: src/cli/serve.ts [hunk 2]
```diff
@@ -518,26 +603,64 @@ export function buildServeRuntime(seam: {
...
+    // settle the status reaction: failed beats attention beats done (a failed
+    // ask never reads as a clean question mark), then drop the hourglass.
+    // Two review-caught exceptions: a drain-presumed-killed turn (same
+    // predicate as runTurn's marker retention) keeps its hourglass - it will
+    // auto-resume next start, and a terminal x nothing ever removes would
+    // read a later successful resume as failed; and a finished thread settles
+    // as done even when the model also flagged attention, because the record
+    // deletion below makes the question mark unremovable forever.
+    const killedByDrain = draining && outcome.failed && !outcome.announcedDrop && !outcome.resultReceived;
+    if (input.messageId && !killedByDrain) {
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/cli/serve.ts]: Applied in 91ab6b0 with your comment wording adopted nearly verbatim. Thanks.

File: src/cli/serve.ts [hunk 3]
```diff
@@ -766,7 +1005,7 @@ export function buildServeRuntime(seam: {
         await thread.post(decision.notice);
         const startedAt = Date.now();
         const outcome = await runTurn({ thread, record: fresh, prompt: decision.prompt, requesterIds, sessionId: decision.sessionId, marker: decision.marker, link });
-        await settleTurn({ thread, outcome, startedAt });
+        await settleTurn({ thread, outcome, startedAt, messageId: decision.marker.messageId, requesterIds });
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/cli/serve.ts]: Agreed, fixed in 91ab6b0: the activeTurn marker now carries the triggering turn's requesterIds, and recovery prefers marker.requesterIds over the streamable handle's newest-author derivation (which stays as the fallback for older markers without the field). Pinned by a recovery test asserting the persisted attention carries the killed turn's askers.

File: src/cli/serve.ts [hunk 4]
```diff
@@ -634,11 +757,122 @@ export function buildServeRuntime(seam: {
...
+   *  outsiders-never-reach-claude invariant covers context lines too), and
+   *  only structurally valid emoji names are ever folded. Removals and
+   *  untracked threads are ignored, as are other bots; our own reactions
+   *  never arrive (chat core drops isMe reaction events before routing). */
+  const onReaction = async (
+    input: { threadId: string; messageId: string; emoji: string; userId: string; isBot?: boolean | "unknown"; added: boolean; occurredAt: number | null },
+    streamable: (threadId: string) => Promise<{ thread: ServeThread }>,
+  ) => {
+    if (!input.added || input.isBot === true) return;
+    if (!loadSlackThread(input.threadId)) return; // untracked thread
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/cli/serve.ts:810]: Agreed, fixed in 91ab6b0: onReaction now drops reactions in unlinked channels up front (serve.reaction_dropped, reason unlinked-channel), before either the answer path or the note path can act. Pinned by a test asserting nothing lands in pendingReactions.

File: src/cli/serve.ts [hunk 5]
```diff
@@ -347,6 +399,10 @@ export function buildServeRuntime(seam: {
   }): Promise<TurnOutcome> => {
     let record: SlackThread = { ...input.record, activeTurn: input.marker };
     saveSlackThread(record);
+    // the whole turn (parks and retries included) reads as "being processed".
+    if (input.marker.messageId) {
+      await setStatus({ threadId: input.thread.id, messageId: input.marker.messageId, emoji: STATUS_EMOJI.processing, op: "add" });
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply src/cli/serve.ts:404]: Agreed for the unknown-recovery drop, fixed in 506bd5a: the terminal drop now settles x and removes the hourglass before returning, mirroring give-up (which already settled). The UNLINKED abandon stays reactionless deliberately: unlinked channels are contractually silent and untouchable (every outbound path honors that), so a leftover hourglass in a channel the owner unlinked is the accepted cost of the contract. Pinned by the drop test asserting the settle.

File: docs/content/docs/serve.es.mdx [hunk 6]
```diff
@@ -17,6 +17,8 @@ Comportamiento destacado:
...
+- **Reacciones de estado**: el mensaje que disparó un turno lleva su estado como una reacción de emoji: un reloj de arena mientras el turno se ejecuta, y luego una marca de verificación (hecho), una x (fallido) o un signo de interrogación (el modelo necesita tu decisión). Requiere el scope `reactions:write` del manifiesto actual; las apps existentes deben volver a pegar el manifiesto y reinstalar.
+- **Atención y avisos**: cuando el modelo te pide una decisión llama a una herramienta en proceso `need_attention`, de modo que el daemon sabe que el hilo te está esperando. Si te quedas en silencio durante 10 minutos, te etiqueta una vez en el hilo. Mientras un hilo está esperando, reaccionar a la pregunta se retransmite como tu respuesta (un pulgar arriba elige la opción recomendada); cualquier otra reacción que añadas en un hilo rastreado se incorpora al siguiente turno como contexto. Requiere `reactions:read` y el evento `reaction_added`.
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply docs/content/docs/serve.es.mdx]: Agreed, fixed in 506bd5a.

File: docs/content/docs/serve.mdx [hunk 7]
```diff
@@ -17,6 +17,8 @@ Notable behavior:
...
+- **Status reactions**: the message that triggered a turn carries its state as an emoji reaction: an hourglass while the turn runs, then a check mark (done), an x (failed), or a question mark (the model needs your decision). Requires the `reactions:write` scope from the current manifest; existing apps must re-paste the manifest and reinstall.
+- **Attention and nudges**: when the model asks you for a decision it calls an in-process `need_attention` tool, so the daemon knows the thread is waiting on you. If you stay quiet for 10 minutes it tags you once in the thread. While a thread is waiting, reacting to the question relays as your answer (a thumbs up picks the recommended option); any other reaction you add in a tracked thread is folded into the next turn as context. Requires `reactions:read` and the `reaction_added` event.
```
[the remark this answers is not in the document]
Comment (anaclumos) [reply docs/content/docs/serve.mdx]: Agreed, fixed in 506bd5a: the attention bullet now carries the same manifest re-paste + reinstall note.
